internal/controller/management/projects/projects.go

```diff
@@ -392,6 +400,74 @@ func (r *reconciler) ensureAPIAdminPermissions(
 	return nil
 }
 
+func (r *reconciler) ensureControllerPermissions(
+	ctx context.Context,
+	project *kargoapi.Project,
+) error {
+	const roleBindingName = "kargo-controller-secrets-readonly"
```

Comment:
Interesting. In my head, each Project was going to have a RoleBinding per controller SA.

You're actually creating or updating a single RoleBinding and making sure it has _all_ the controller SAs listed as its subjects.

On some level, I quite like that, but I have a worry this will be more complicated than anticipated once you get into part two of this PR...

Remember that we also need to deal with creating/removing bindings as new controller SAs come and go. Once we start doing that, you start to run into possibilities like a new Project and a new SA being reconciled concurrently and racing to update the same binding.

One tempting solution to that problem is to _patch_ the RoleBinding instead of updating it, but patching _lists_ can be quite difficult.

In the end, it may be more straightforward if every Project has one RoleBinding _per controller SA._ I feel that would lead to fewer conflicts and race conditions.

But I also want @hiddeco's opinion on this matter.

Reply:
Note that apart from this, this PR is off to a very good start.

Reply:
Let's wait to hear from @hiddeco before making any changes. He usually has the best insight into avoiding races in controllers.

Reply:
@hiddeco any opinion on just using a separate RoleBinding per SA/Project pair? I cannot see a race that way and wonder if it isn't simpler than your option 1.

Reply:
Sorry. Brain is still booting this AM...

> You put one reconciler in control of the reconciliation of the RoleBinding

ikd if this is a viable option. It's not the RoleBindings that need to be reconciled.

RoleBindings need to be updated _or_ created/removed in response to both Projects and controller SAs being added/removed.

Reply:
You can create a reconciler around an arbitrary resource type since controller-runtime v0.19.0. See:

- https://github.com/kubernetes-sigs/controller-runtime/blob/main/examples/typed/main.go
- https://github.com/kubernetes-sigs/controller-runtime/blob/main/examples/multiclustersync/main.go